lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -5919,50 +5919,60 @@ LogicalResult LayerBlockOp::verify() {
   }
 
   // Verify the body of the region.
-  Block *body = getBody(0);
-  bool failed = false;
-  body->walk<mlir::WalkOrder::PreOrder>([&](Operation *op) {
-    // Skip nested layer blocks.  Those will be verified separately.
-    if (isa<LayerBlockOp>(op))
-      return WalkResult::skip();
-    // Check all the operands of each op to make sure that only legal things are
-    // captured.
-    for (auto operand : op->getOperands()) {
-      // Any value captured from the current layer block is fine.
-      if (operand.getParentBlock() == body)
-        continue;
-      // Capture of a non-base type, e.g., reference, is allowed.
-      FIRRTLBaseType baseType = dyn_cast<FIRRTLBaseType>(operand.getType());
-      if (!baseType)
-        return WalkResult::advance();
-      // Capturing a non-passive type is illegal.
-      if (!baseType.isPassive()) {
-        auto diag = emitOpError()
-                    << "captures an operand which is not a passive type";
-        diag.attachNote(operand.getLoc()) << "operand is defined here";
-        diag.attachNote(op->getLoc()) << "operand is used here";
-        failed = true;
-        return WalkResult::advance();
-      }
-    }
-    // Ensure that the layer block does not drive any sinks.
-    if (auto connect = dyn_cast<FConnectLike>(op)) {
-      auto dest = getFieldRefFromValue(connect.getDest()).getValue();
-      if (dest.getParentBlock() == body)
+  auto result = getBody(0)->walk<mlir::WalkOrder::PreOrder>(
+      [&](Operation *op) -> WalkResult {
+        // Skip nested layer blocks.  Those will be verified separately.
+        if (isa<LayerBlockOp>(op))
+          return WalkResult::skip();
+
+        // Check all the operands of each op to make sure that only legal things
+        // are captured.
+        for (auto operand : op->getOperands()) {
+          // Any value captured from the current layer block is fine.
+          if (auto *definingOp = operand.getDefiningOp())
+            if (getOperation()->isAncestor(definingOp))
+              continue;
+
+          // Capture of a non-base type, e.g., reference, is allowed.
+          FIRRTLBaseType type = dyn_cast<FIRRTLBaseType>(operand.getType());
+          if (!type)
+            continue;
+
+          // Capturing a non-passive type is illegal.
+          if (!type.isPassive()) {
+            auto diag = emitOpError()
+                        << "captures an operand which is not a passive type";
+            diag.attachNote(operand.getLoc()) << "operand is defined here";
```

Comment:
`emitOpError()` will identify the layerblock as the source of the error, I think that suffices, yeah? 

Also, the verifier for the connect op runs first, and gives a slick error message:
```
FIRRTL version 4.0.0

circuit Top:
  layer A, bind:
  module Top:
    wire w : UInt<1>
    layerblock A:
      connect w, UInt<1>(0)
```

```
../scratch.fir:8:7: error: 'firrtl.strictconnect' op connects to a destination which is defined outside its enclosing layer block
      connect w, UInt<1>(0)
      ^
../scratch.fir:8:7: note: see current operation: "firrtl.strictconnect"(%0, %2) : (!firrtl.uint<1>, !firrtl.uint<1>) -> ()
../scratch.fir:7:5: note: enclosing layer block is defined here
    layerblock A:
    ^
../scratch.fir:6:5: note: destination is defined here
    wire w : UInt<1>
    ```

Reply:
I don't know how the connect verifier is related to this one but I believe you :+1: .

FWIW the connect error uses `emitOpError()` but takes care to do:
```
          diag.attachNote(getLoc()) << "enclosing layer block is defined here";
```

Which is partly what made me think about this here.

Does `emitOpError()` do something reasonable with the `-mlir-print-op-on-diagnostic=0` option?

Anyway that is slick, awesome!

Reply:
> I don't know how the connect verifier is related to this one 

Just saying that, if there is a connect under a layerblock, the verifier for the connect will run first (because verifiers run bottom up), and since that verifier already checks that the destination of the connect is within the same layerblock as the connect itself, it's actually impossible to produce the error message we are discussing.

> Does emitOpError() do something reasonable with the -mlir-print-op-on-diagnostic=0 option?

Hmm! I didn't know about that option! Yes it seems to be just fine:
```
./bin/firtool --parse-only  ../scratch.fir --mlir-print-ir-after-all -mlir-print-op-on-diagnostic=0
../scratch.fir:8:7: error: 'firrtl.strictconnect' op connects to a destination which is defined outside its enclosing layer block
      connect w, UInt<1>(0)
      ^
../scratch.fir:7:5: note: enclosing layer block is defined here
    layerblock A:
    ^
../scratch.fir:6:5: note: destination is defined here
    wire w : UInt<1>
```

Reply:
Oh ignore me, I am wrong about where the error came from